middleware.go

```diff
@@ -0,0 +1,679 @@
+	spec.Parameters = mergeParameters(spec.Parameters, p.defaults.parameters())
+	return spec, nil
+}
+
+func (p *processorWithSchemaDecode) Configure(ctx context.Context, config config.Config) error {
+	err := p.Processor.Configure(ctx, config)
+	if err != nil {
+		return err //nolint:wrapcheck // middleware shouldn't wrap errors
+	}
+
+	p.keyEnabled = *p.defaults.KeyEnabled
+	if val, ok := config[configProcessorWithSchemaDecodeKeyEnabled]; ok {
+		p.keyEnabled, err = strconv.ParseBool(val)
+		if err != nil {
+			return fmt.Errorf("invalid %s: failed to parse boolean: %w", configProcessorWithSchemaDecodeKeyEnabled, err)
+		}
+	}
+
+	p.payloadEnabled = *p.defaults.PayloadEnabled
+	if val, ok := config[configProcessorWithSchemaDecodePayloadEnabled]; ok {
+		p.payloadEnabled, err = strconv.ParseBool(val)
+		if err != nil {
+			return fmt.Errorf("invalid %s: failed to parse boolean: %w", configProcessorWithSchemaDecodePayloadEnabled, err)
+		}
+	}
+
+	return nil
+}
+
+func (p *processorWithSchemaDecode) Process(ctx context.Context, records []opencdc.Record) []ProcessedRecord {
+	if p.keyEnabled {
+		for i := range records {
+			if err := p.decodeKey(ctx, &records[i]); err != nil {
+				if len(records) > 0 {
+					err = fmt.Errorf("record %d: %w", i, err)
+				}
+				return []ProcessedRecord{ErrorRecord{Error: err}}
+			}
+		}
+	}
+	if p.payloadEnabled {
+		for i := range records {
+			if err := p.decodePayload(ctx, &records[i]); err != nil {
+				if len(records) > 0 {
+					err = fmt.Errorf("record %d: %w", i, err)
+				}
+				return []ProcessedRecord{ErrorRecord{Error: err}}
+			}
+		}
+	}
+
+	return p.Processor.Process(ctx, records)
+}
+
+func (p *processorWithSchemaDecode) decodeKey(ctx context.Context, rec *opencdc.Record) error {
+	subject, errSubject := rec.Metadata.GetKeySchemaSubject()
+	version, errVersion := rec.Metadata.GetKeySchemaVersion()
+	switch {
+	case errSubject != nil && !errors.Is(errSubject, opencdc.ErrMetadataFieldNotFound):
+		return fmt.Errorf("failed to get key schema subject from metadata: %w", errSubject)
+	case errVersion != nil && !errors.Is(errVersion, opencdc.ErrMetadataFieldNotFound):
+		return fmt.Errorf("failed to get key schema version from metadata: %w", errVersion)
+	case errors.Is(errSubject, opencdc.ErrMetadataFieldNotFound) ||
+		errors.Is(errVersion, opencdc.ErrMetadataFieldNotFound):
+		// log warning once, to avoid spamming the logs
+		p.keyWarnOnce.Do(func() {
+			Logger(ctx).Warn().Msgf(`record does not have an attached schema for the key, consider disabling the processor schema key decoding using "%s: false"`, configProcessorWithSchemaDecodeKeyEnabled)
+		})
+		return nil
+	}
+
+	if rec.Key != nil {
+		decodedKey, err := p.decode(ctx, rec.Key, subject, version)
+		if err != nil {
+			return fmt.Errorf("failed to decode key: %w", err)
+		}
+		rec.Key = decodedKey
+	}
+
+	return nil
+}
+
+func (p *processorWithSchemaDecode) decodePayload(ctx context.Context, rec *opencdc.Record) error {
```

Comment:
this method and `decodeKey` are almost identical, we can have the similarities in another function to not have duplicates, WDYT?
ofc not a blocker

Reply:
Yeah, they are awfully similar, but have differences in key places, like taking stuff from the metadata, taking data from the record or the error messages. Also the fact that the payload has 2 subfields (before and after) doesn't make things easier. I was thinking about extracting it, but I feel like it would have a negative impact on the readability of the code 😕